### PR TITLE
UAR:433 - Summary view new BO's

### PR DESCRIPTION
**Pull request by stephanie-ch**
### JIRA link
https://companieshouse.atlassian.net/browse/UAR-433

### Change description
Add summary table of newly added BO's

### Work checklist

- [ ] Tests added where applicable
- [ ] UI changes meet accessibility criteria

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: views/update/update-beneficial-owner-type.html [hunk 1]
```diff
@@ -56,6 +56,73 @@
...
+          attributes: {
+            "data-event-id": "what-is-a-managing-officer-details"
+          },
+          html: "<p>This is a director, manager or company secretary of the overseas entity.</p>"
+        }) }}
+      {% endif %}
+
+      <span class="govuk-caption-xl govuk-!-padding-bottom-1">{{ entity_name }} - {{ entity_number }}</span>
+
+</div>
```
Comment (stephanie-ch) [inline views/update/update-beneficial-owner-type.html:81]: Will tidy this file, aware the indentation is off

File: src/controllers/update/beneficial.owner.type.controller.ts [hunk 2]
```diff
@@ -33,7 +33,7 @@ export const get = async (req: Request, res: Response, next: NextFunction) => {
       backLinkUrl: config.UPDATE_BENEFICIAL_OWNER_BO_MO_REVIEW_URL,
       templateName: config.UPDATE_BENEFICIAL_OWNER_TYPE_PAGE,
       ...appData,
-      noLists: true
+      noLists: false
```
Comment (AndreKramer7654) [inline src/controllers/update/beneficial.owner.type.controller.ts:36]: maybe we don't actually need this config now

File: src/utils/beneficial.owner.gov.ts [hunk 3]
```diff
@@ -127,6 +140,13 @@ const setBeneficialOwnerData = (reqBody: any, id: string): ApplicationDataType =
     : {};
   data[StartDateKey] = mapFieldsToDataObject(reqBody, StartDateKeys, InputDateKeys);
 
+  // not present in register journey
```
Comment (AndreKramer7654) [inline src/utils/beneficial.owner.gov.ts:143]: comment not needed / confusing to me

File: src/utils/beneficial.owner.other.ts [hunk 4]
```diff
@@ -129,6 +142,13 @@ const setBeneficialOwnerData = (reqBody: any, id: string): ApplicationDataType =
     : {};
   data[StartDateKey] = mapFieldsToDataObject(reqBody, StartDateKeys, InputDateKeys);
 
+  // not present in register journey
+  if ("is_still_bo" in reqBody) {
+    data[CeasedDateKey] = (reqBody["is_still_bo"] === '0')
```
Comment (AndreKramer7654) [inline src/utils/beneficial.owner.other.ts:147]: also candidate for shared method

File: views/includes/list/update-beneficial-owner.html [hunk 5]
```diff
@@ -0,0 +1,64 @@
+{% if hasBeneficialOwners %}
+    <table class="govuk-table">
+        <caption class="govuk-table__caption govuk-table__caption--m">What you have added in this update</caption>
```
[the remark this answers is not in the document]
Comment (AndreKramer7654) [reply views/includes/list/update-beneficial-owner.html:3]: or better make it re-usable and share for review list too? If that is possible!

File: src/utils/beneficial.owner.other.ts (hunk 4 above)
Comment (AndreKramer7654) [reply src/utils/beneficial.owner.other.ts:147]: in my PR I've just done the === '0' as anything else means not ceased. So no need for common methods as just one line.

File: src/utils/beneficial.owner.gov.ts (hunk 3 above)
Comment (stephanie-ch) [reply src/utils/beneficial.owner.gov.ts:143]: removed

File: src/utils/beneficial.owner.gov.ts [hunk 6]
```diff
@@ -39,20 +42,30 @@ export const getBeneficialOwnerGovById = (req: Request, res: Response, next: Nex
...
       templateName: `${templateName}/${id}`,
       id,
       ...data,
       ...principalAddress,
       ...serviceAddress,
       [StartDateKey]: startDate
-    });
+    };
+
+    // extra data needed for update journey
```
[the remark this answers is not in the document]
Comment (stephanie-ch) [reply src/utils/beneficial.owner.gov.ts:61]: i've removed the comment, as it's now in a function

File: src/utils/beneficial.owner.other.ts [hunk 7]
```diff
@@ -41,20 +44,30 @@ export const getBeneficialOwnerOtherById = (req: Request, res: Response, next: N
...
       ...data,
       ...principalAddress,
       ...serviceAddress,
       [StartDateKey]: startDate
-    });
+    };
+
+    // extra data needed for update journey
+    if (EntityNumberKey in appData && appData[EntityNumberKey] !== null) {
+      templateOptions["is_still_bo"] = (Object.keys(data["ceased_date"]).length === 0) ? 1 : 0;
```
[the remark this answers is not in the document]
Comment (stephanie-ch) [reply src/utils/beneficial.owner.other.ts:65]: shared function created

File: views/includes/list/update-beneficial-owner.html (hunk 5 above)
[the remark this answers is not in the document]
Comment (stephanie-ch) [reply views/includes/list/update-beneficial-owner.html:3]: I've renamed the file. Il see what Christy has done for his list and perhaps we can combine them in his ticket.

File: src/utils/beneficial.owner.gov.ts [hunk 8]
```diff
@@ -39,20 +42,30 @@ export const getBeneficialOwnerGovById = (req: Request, res: Response, next: Nex
...
+    if (EntityNumberKey in appData && appData[EntityNumberKey] !== null) {
```
Comment (AndreKramer7654) [inline src/utils/beneficial.owner.gov.ts:62]: is null the right thing here to check for?

File: src/controllers/update/beneficial.owner.type.controller.ts [hunk 9]
```diff
@@ -158,3 +164,19 @@ export const retrieveBeneficialOwners = async (req: Request, appData: Applicatio
     }
   }
 };
+
+const checkForNewlyAddedBos = (appData: ApplicationData) => {
+  const allBos = [
```
Comment (AndreKramer7654) [inline src/controllers/update/beneficial.owner.type.controller.ts:169]: quite expensive to concat all the lists just to check all as one

**Review by AndreKramer7654 — APPROVED** [review]

File: src/controllers/update/beneficial.owner.type.controller.ts (hunk 9 above)
Comment (stephanie-ch) [reply src/controllers/update/beneficial.owner.type.controller.ts:169]: Yeah - any suggestions? Other options I tried weren't safe as the lists are optional

**Review by corybudd-ch — APPROVED** [review]
Duplication sitting at 5% at the minute but after looking at it, difficult to reduce this below 3% due to the nature of existing code and refactoring too much of registration code to satisfy Sonar's duplication metrics by another 2%. So, happy to approve and would be happy for it to still be merged

**Comment by AndreKramer7654** [comment]
> Duplication sitting at 5% at the minute but after looking at it, difficult to reduce this below 3% due to the nature of existing code and refactoring too much of registration code to satisfy Sonar's duplication metrics by another 2%. So, happy to approve and would be happy for it to still be merged

It's estimating 1.2% duplication after merge so I agree.

File: src/controllers/update/beneficial.owner.type.controller.ts (hunk 9 above)
Comment (AndreKramer7654) [reply src/controllers/update/beneficial.owner.type.controller.ts:169]: You could process each loop separately.

Comment (AndreKramer7654) [reply src/controllers/update/beneficial.owner.type.controller.ts:169]: Not tried this but how about something like (not actually tried this):
const hasReference = bo => bo.ch_reference === undefined;
return (appData[BeneficialOwnerIndividualKey] ?? []).find(hasReference) === true || (appData[BeneficialOwnerOtherKey] ?? []).find(hasReference) === true || (appData[BeneficialOwnerGovKey] ?? []).find(hasReference) === true;

Comment (stephanie-ch) [reply src/controllers/update/beneficial.owner.type.controller.ts:169]: Tried suggestion but it gives a type mismatch between BOType/undefined and boolean.

**Pull request merged**